app/code/Magento/SalesRule/Model/Validator.php

```diff
@@ -385,6 +386,12 @@ public function processShippingAmount(Address $address)
                     }
                     $address->setCartFixedRules($cartRules);
                     break;
+                case Rule::BUY_X_GET_Y_ACTION:
+                    $allQtyDiscount = $this->getDiscountQtyAllItemsBuyXGetYAction($quote, $rule);
```

Comment:
Seems not need apply discount to Shipping, 
if Shipping method Handling type is "Per Order" or "Fixed" (fixed shipping price for Order).

![Configuration___Settings___Stores___Magento_Admin](https://user-images.githubusercontent.com/60192090/85866072-81ae0480-b7cf-11ea-9b92-00c33e912cd0.jpg)
![Shopping_Cart](https://user-images.githubusercontent.com/60192090/85866082-84a8f500-b7cf-11ea-8505-1de29fc41483.jpg)

Reply:
Checked
Your changes are correct.